### DISCUSSION TITLE
Problem with Contract command

### DISCUSSION THREAD

**Discussion by jdcastanoy** [FeynCalc/feyncalc#125] [Q&A]
I'm working with the following tensor:
`B[q_, a_, b_] := 
 Pair[LorentzIndex[a], LorentzIndex[b]] - (
  Pair[Momentum[q], LorentzIndex[a]] Pair[Momentum[q], 
    LorentzIndex[b]])/
  Contract[Pair[Momentum[q], LorentzIndex[a]] Pair[Momentum[q], 
     LorentzIndex[a]]]`

with the following restriction:

`Contract[Pair[LorentzIndex[a], LorentzIndex[b]] Pair[LorentzIndex[a], 
     LorentzIndex[b]]] = 2;`

Now, the contraction of the tensor with itself is 1 (it can be easely checked), but when I try to compute it with Feyncalc (I will need more complicated contractions in the future) the result is 3:

In[]: `Contract[B[q, a, b] B[q, a, b]]`
Out[] : 3

Nevertheless, if I do the contractions term by term, all seems work. The problem is when I put the whole tensor inside `Contract`.
Can you help me with this? 

I have FeynCalcl 9.3.1 (stable version) which I use in Mathematica 11.2 Student Edition for windows

Thanks.

**Comment by vsht** [comment]
I think that this problem is related to #95. Currently you can have only 3+1 or D dimensions, 
where D around 4. It is not directly possible to work in 1+1, 2+1 etc. The type of constraints
your are trying to set are equally not supported.

You can try to calculate your tensor in D dimensions and set D->2 at the end. This should work for
vectors and metric tensors, but will of course fail for the Levi-Civita. Of course, one has to check
that everything works out and there are no unexpected side effects.

**Reply by jdcastanoy** [reply]
Thanks for your response.
Nevertheless, even without the condition over the metric contraction, the result should be 1, but the command gives me 3.
I don't know if is because I installed a new version with a bug or is my Mathematica, because in the past such contractions worked well.

**Reply by vsht** [reply]
>  Nevertheless, even without the condition over the metric contraction, the result should be 1, but the command gives me 3.

Sorry, but I don't understand this statement. The result of the contraction obviously depends on the number of the dimensions.
The obvious interpretation for me is that 3 is the number of polarizations of a massive vector meson in 4 dimensions, cf.
`PolarizationSum[mu, nu, q]`. Now

```
B[q_, a_, b_] := 
 Pair[LorentzIndex[a], 
   LorentzIndex[
    b]] - (Pair[Momentum[q], LorentzIndex[a]] Pair[Momentum[q], 
      LorentzIndex[b]])/
   Contract[
    Pair[Momentum[q], LorentzIndex[a]] Pair[Momentum[q], 
      LorentzIndex[a]]]
```

```
{B[q, a, b][[1]] B[q, a, b][[1]],
 B[q, a, b][[1]] B[q, a, b][[2]],
 B[q, a, b][[2]] B[q, a, b][[1]],
 B[q, a, b][[2]] B[q, a, b][[2]]}
% // Contract
(*{4,-1,-1,1}*)
```

If the first contraction gives 2, then the sum is 1 as you say, but not otherwise. Besides,

```
{B[q, a, b][[1]] B[q, a, b][[1]],
 B[q, a, b][[1]] B[q, a, b][[2]],
 B[q, a, b][[2]] B[q, a, b][[1]],
 B[q, a, b][[2]] B[q, a, b][[2]]}
% // ChangeDimension[#, D] & // Contract
% /. D -> 2
(*{2,-1,-1,1}*)
```
works exactly as expected. So I somehow fail to see what is the problem with this behavior.

> I don't know if is because I installed a new version with a bug or is my Mathematica, because in the past such contractions worked well.

Can you give some examples of what worked well in the past and with which versions?
